iree/compiler/Dialect/HAL/IR/test/command_buffer_ops.mlir

```diff
@@ -1,137 +1,187 @@
-// Tests printing and parsing of hal.command_buffer ops.
-
-// RUN: iree-opt -allow-unregistered-dialect -split-input-file %s | iree-opt -allow-unregistered-dialect -split-input-file | IreeFileCheck %s
-
-// CHECK-LABEL: @make_memory_barrier
-func @make_memory_barrier() -> tuple<i32, i32> {
-  // CHECK: %memory_barrier = hal.make_memory_barrier "HostRead|HostWrite", "MemoryRead|MemoryWrite" : tuple<i32, i32>
-  %memory_barrier = hal.make_memory_barrier "HostRead|HostWrite", "MemoryRead|MemoryWrite" : tuple<i32, i32>
-  return %memory_barrier : tuple<i32, i32>
-}
-
-// -----
-
-// CHECK-LABEL: @make_buffer_barrier
-func @make_buffer_barrier(%arg0 : !hal.buffer) -> tuple<i32, i32, !hal.buffer, index, index> {
-  %0 = "test_hal.offset"() : () -> index
-  %1 = "test_hal.length"() : () -> index
-  // CHECK: %buffer_barrier = hal.make_buffer_barrier "HostRead|HostWrite", "MemoryRead|MemoryWrite", %arg0, %0, %1 : tuple<i32, i32, !hal.buffer, index, index>
-  %buffer_barrier = hal.make_buffer_barrier "HostRead|HostWrite", "MemoryRead|MemoryWrite", %arg0, %0, %1 : tuple<i32, i32, !hal.buffer, index, index>
-  return %buffer_barrier : tuple<i32, i32, !hal.buffer, index, index>
-}
-
-// -----
+// RUN: iree-opt -split-input-file %s | iree-opt -split-input-file | IreeFileCheck %s
 
 // CHECK-LABEL: @command_buffer_create
-func @command_buffer_create(%arg0 : !hal.device) {
-  // CHECK: %cmd = hal.command_buffer.create %arg0, OneShot, "Transfer|Dispatch" : !hal.command_buffer
-  %cmd = hal.command_buffer.create %arg0, OneShot, "Transfer|Dispatch" : !hal.command_buffer
+//  CHECK-SAME: (%[[DEVICE:.+]]: !hal.device)
+func @command_buffer_create(%device: !hal.device) {
+  //      CHECK: %cmd = hal.command_buffer.create
+  // CHECK-SAME:   device(%[[DEVICE]] : !hal.device)
+  // CHECK-SAME:   mode(OneShot)
+  // CHECK-SAME:   categories("Transfer|Dispatch") : !hal.command_buffer
+  %cmd = hal.command_buffer.create device(%device : !hal.device)
+                                     mode(OneShot)
+                               categories("Transfer|Dispatch") : !hal.command_buffer
```

Comment:
Hm.. I'm wondering about the syntax (and printer/parser) of this op, and others like `execution_barrier`. Typically there would be commas or some other separator between arguments, instead of newlines, and some wrapper around the parameters like `[  ]`, `(  )`, `{  }`, `<  >`. This certainly reads easily as a human though. WDYT?

Reply:
the newlines are optional (it just needs whitespace) - the syntax here matches linalg which has things like:
```mlir
  linalg.generic #attrs
    ins(%A, %B: memref<?x?xf32>, memref<?x?xf32>)
    outs(%C: memref<?x?xf32>) {
    ^bb0(%a: f32, %b: f32, %c: f32):
      %d = addf %a, %b : f32
      linalg.yield %d : f32
    }
```

The commas and simple separators have ambiguity when parsing the nested types (which is why linalg does this too). Future PRs will change some of these enums to being type parameters instead, so the op will be `%cmd = hal.command_buffer.create device(%device : !hal.device) : !hal.command_buffer<..., "OneShot", "Transfer|Dispatch">` (or something like that). Same with memory types and such for buffers - those will be on the return types instead of as op attributes.